# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: extensions/permissions/PermissionDelegateHandler.cpp
# Commit: 08c6683cd6be
# Full Hash: 08c6683cd6bec0d7ad40f6fef30460484d7f51e0
# Author: Tim Huang <tihuang@mozilla.com>
# Date: 2020-06-22 21:47:58
# Description:
#   Bug 1645997 - Use BrowsingContext to get the top-level window context in PermissionDelegateHandler.cpp r=baku
#   
#   We used the Document::GetWindowContext() to get the window context
#   related to the doucment. But, this could return a nullptr if the
#   document is detached from the window or the docuemnt is destoried.
# ==============================================================================

diff -r aa66effc4d07 -r 08c6683cd6be extensions/permissions/PermissionDelegateHandler.cpp
--- a/extensions/permissions/PermissionDelegateHandler.cpp	Mon Jun 22 10:36:38 2020 +0000
+++ b/extensions/permissions/PermissionDelegateHandler.cpp	Fri Jun 19 12:31:29 2020 +0000
@@ -154,16 +154,20 @@
 static bool IsCrossOriginContentToTop(Document* aDocument) {
   MOZ_ASSERT(aDocument);
 
-  BrowsingContext* topBrowsingContext = aDocument->GetBrowsingContext()->Top();
+  RefPtr<BrowsingContext> bc = aDocument->GetBrowsingContext();
+  if (!bc) {
+    return true;
+  }
+  RefPtr<BrowsingContext> topBC = bc->Top();
 
   // In Fission, we can know if it is cross-origin by checking whether both
   // contexts in the same process. So, If they are not in the same process, we
   // can say that it's cross-origin.
-  if (!topBrowsingContext->IsInProcess()) {
+  if (!topBC->IsInProcess()) {
     return true;
   }
 
-  RefPtr<Document> topDoc = topBrowsingContext->GetDocument();
+  RefPtr<Document> topDoc = topBC->GetDocument();
   if (!topDoc) {
     return true;
   }
@@ -248,11 +252,15 @@
   }
 
   nsIPrincipal* principal = mPrincipal;
+  // If we cannot get the browsing context from the document, we fallback to use
+  // the prinicpal of the document to test the permission.
+  RefPtr<BrowsingContext> bc = mDocument->GetBrowsingContext();
+
   if ((info->mPolicy == DelegatePolicy::eDelegateUseTopOrigin ||
        (info->mPolicy == DelegatePolicy::eDelegateUseFeaturePolicy &&
-        StaticPrefs::dom_security_featurePolicy_enabled()))) {
-    RefPtr<WindowContext> topWC =
-        mDocument->GetWindowContext()->TopWindowContext();
+        StaticPrefs::dom_security_featurePolicy_enabled())) &&
+      bc) {
+    RefPtr<WindowContext> topWC = bc->GetTopWindowContext();
 
     if (topWC->IsInProcess()) {
       // If the top-level window context is in the same process, we directly get
